ksql-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -204,23 +210,33 @@ private ProjectNode buildProjectNode(final PlanNode sourcePlanNode) {
 
     final Optional<ColumnName> keyFieldName = getSelectAliasMatching((expression, alias) ->
         expression instanceof ColumnReferenceExp
-            && ((ColumnReferenceExp) expression).getReference().equals(sourceKeyFieldName)
+            && ((ColumnReferenceExp) expression).getReference().equals(sourceKeyFieldName),
+        sourcePlanNode
     );
 
     return new ProjectNode(
         new PlanNodeId("Project"),
         sourcePlanNode,
         schema,
         keyFieldName.map(ColumnRef::withoutSource),
-        analysis.getSelectExpressions()
+        sourcePlanNode.getSelectExpressions()
     );
   }
 
   private static FilterNode buildFilterNode(
       final PlanNode sourcePlanNode,
       final Expression filterExpression
   ) {
-    return new FilterNode(new PlanNodeId("Filter"), sourcePlanNode, filterExpression);
+    return new FilterNode(new PlanNodeId("Filter"), sourcePlanNode, filterExpression,
+        sourcePlanNode.getSelectExpressions());
+  }
+
+  private FlatMapNode buildFlatMapNode(
+      final PlanNode sourcePlanNode,
+      final FunctionRegistry functionRegistry
+  ) {
+    return new FlatMapNode(new PlanNodeId("FlatMap"), sourcePlanNode,
+        sourcePlanNode.getSchema(), functionRegistry, analysis);
```

Comment:
you can get `sourcePlanNode.getSchema()` from `sourcePlanNode`, which is also passed to the constructor. So no need to pass it separately.

Reply:
done